biorbd_optim/__init__.py

```diff
@@ -134,34 +135,37 @@ def __init__(
         self.g = []
         self.g_bounds = Bounds()
         Constraint.continuity_constraint(self)
-
-        if self.nb_phases == 1:
-            if len(constraints) == 0 or (
-                isinstance(constraints, (list, tuple))
-                and isinstance(constraints[0], (list, tuple))
-                and not isinstance(constraints[0][0], (list, tuple))
-            ):
-                constraints = (constraints,)
-        self.__add_to_nlp("constraints", constraints, False)
-        for i in range(self.nb_phases):
-            Constraint.add_constraints(self, self.nlp[i])
+        if len(constraints) > 0:
+            if self.nb_phases == 1:
+                if isinstance(constraints, dict):
+                    constraints = (constraints,)
+                if isinstance(constraints[0], dict):
+                    constraints = (constraints,)
+            elif isinstance(constraints, (list, tuple)):
+                constraints = list(constraints)
+                for i, constraint in enumerate(constraints):
+                    if isinstance(constraint, dict):
+                        constraints[i] = (constraint,)
+            self.__add_to_nlp("constraints", constraints, False)
+            for i in range(self.nb_phases):
+                Constraint.add_constraints(self, self.nlp[i])
 
         # Objective functions
         self.J = 0
-        if self.nb_phases == 1:
-            if (
-                isinstance(objective_functions, (list, tuple))
-                and isinstance(objective_functions[0], (list, tuple))
-                and not isinstance(objective_functions[0][0], (list, tuple))
-            ):
-                objective_functions = (objective_functions,)
-        self.__add_to_nlp("objective_functions", objective_functions, False)
-        for i in range(self.nb_phases):
-            for (func, params) in self.nlp[i]["objective_functions"]:
-                if isinstance(params, dict):
-                    func(self, self.nlp[i], **params)
-                else:
-                    func(self, self.nlp[i], params)
+        if len(objective_functions) > 0:
+            if self.nb_phases == 1:
+                if isinstance(objective_functions, dict):
+                    objective_functions = (objective_functions,)
+                if isinstance(objective_functions[0], dict):
+                    objective_functions = (objective_functions,)
+            elif isinstance(objective_functions, (list, tuple)):
```

Comment:
What happens if self.nb_phases > 1 but objective_functions is not list or tuple?

Reply:
error line 197 " raise RuntimeError(param_name + " must be a list or tuple when number of phase is not equal to 1")"

 because 2 phases !